docs(refactor): refactor and refresh of filter content (#1366)
kroxylicious/kroxylicious: docs/modules/record-encryption/con-record-encryption-overview.adoc

// file included in the following:
//
// assembly-record-encryption-filter.adoc

[id='con-topic-encryption-overview-{context}']

> = Encryption process

Reviewer: Maybe we could align it with the other heading `How Kroxylicious works` like`How Record Encryption works`.

[role="_abstract"]
The Record Encryption filter uses envelope encryption to encrypt records with symmetric encryption keys. 
The filter encrypts records from produce requests and decrypts records from fetch responses.

Envelope encryption::
Envelope encryption is an industry-standard technique suited for encrypting large volumes of data in an efficient manner.
Data is encrypted with a Data Encryption Key (DEK).
The DEK is encrypted using a Key Encryption Key (KEK).
The KEK is stored securely in a Key Management System (KMS).
Symmetric encryption keys::
AES(GCM) 256 bit encryption symmetric encryption keys are used to encrypt and decrypt record data. 

The process is as follows:

> . The filter intercepts produce requests from producing applications and encrypts the records.

Reviewer 2: ```suggestion
. The filter intercepts produce requests from producing applications and transforms it by encrypting the records.
```

. The produce request is forwarded to the broker. 

> . The filter intercepts fetch responses from consuming applications and decrypts the records.

Reviewer 2: ```suggestion
. The filter intercepts fetch responses from the broker and transforms it by decrypting the records.
```

. The fetch response is forwarded to the consuming application.

The filter encrypts the record value only.
Record keys, headers, and timestamps are not encrypted.

The entire process is transparent from the point of view of Kafka clients and Kafka brokers. 
Neither are aware that the records are being encrypted, nor do they have any access to the encryption keys or have any influence on the ciphering process to secure the records.

== How the filter encrypts records
The filter encrypts records from produce requests as follows:

. Filter selects a KEK to apply.
. Requests the KMS to generate a DEK for the KEK.
. Uses an encrypted DEK (DEK encrypted with the KEK) to encrypt the record.

> . Replaces the original record with a cipher record (encrypted record, encrypted DEK, and metadata).

Reviewer 2: Is it a `cipher record` or `enciphered record` ? I'm not sure what the correct/ commonly accepted term is.

Reviewer 3: In the design we use the term "ciphertext record" to describe the record content that gets forwarded to the broker.   I think that's the term we should favour in the user facing documentation.

https://github.com/kroxylicious/kroxylicious/blob/4535cc01e7de2face81ea9d65827988f6d1c3a8a/kroxylicious-filters/kroxylicious-record-encryption/doc/design.adoc?plain=1#L231

Reviewer 2: I'm all for consistency so `ciphertext record` would be the path of least resistance but still sounds a bit odd to me. 

The more I think about it the harder time I have coming up with a phrase that works for me, so lets just be consistent unless @PaulRMellor has better idea.

The filter uses a DEK reuse strategy.
Encrypted records are sent to the same topic using the same DEK until a time-out or an encryption limit is reached.

== How the filter decrypts records
The filter decrypts records from fetch responses as follows:

. Filter receives a cipher record from the Kafka broker.
. Reverses the process that constructed the cipher record.
. Uses KMS to decrypt the DEK.
. Uses the decrypted DEK to decrypt the encrypted record.
. Replaces the cipher record with a decrypted record.

> The filter uses an LRU (least recently used) strategy for decrypted records.

Reviewer 2: The filter DOES NOT cache decrypted records, It caches Decrypted DEKS!

Decrypted DEKs are kept in memory to reduce interactions with the KMS.

== How the filter uses the KMS
To support the filter, the KMS provides the following:

* A secure repository for storing Key Encryption Keys (KEKs)
* A service for generating and decrypting Data Encryption Keys (DEKs)

KEKs stay within the KMS.
The KMS generates a DEK (which is securely generated random data) for a given KEK, then returns the DEK and an encrypted DEK.
The encrypted DEK has the same data but encrypted with the KEK.
The KMS doesn't store DEKs; they are stored as part of the cipher record in the broker.

WARNING: The KMS must be available during runtime.

> If the KMS is unavailable, production and consumption through the filter become impossible until KMS service is restored.

Reviewer 2: The DEK caches mitigate this somewhat. Though I'm not sure if we want to call that out.

. Uses the decrypted DEK to decrypt the encrypted record.
. Replaces the cipher record with a decrypted record.

> The filter uses an LRU (least recently used) strategy for caching decrypted DEKS.

Reviewer 3: ```suggestion
The filter uses an LRU (least recently used) strategy for caching decrypted DEKs.
```

KEKs stay within the KMS.
The KMS generates a DEK (which is securely generated random data) for a given KEK, then returns the DEK and an encrypted DEK.
The encrypted DEK has the same data but encrypted with the KEK.
The KMS doesn't store encrypted DEKs; they are stored as part of the cipher record in the broker.

> WARNING: The KMS must be available during runtime. 

Reviewer 3: ```suggestion
WARNING: The KMS must be available during runtime.
```
